File: THCS/2023_2024/TuyenHoa_2324/Cau3/CAKE.py
def main():
    # Đọc dữ liệu từ file
    with open('CAKE.INP', 'r') as infile:
        N = int(infile.readline().strip())
        cake = infile.readline().strip()

    # Đếm tổng số quả dâu và quả sim
    total_D = cake.count('D')
    total_S = cake.count('S')

    # Nếu số lượng quả dâu hoặc sim là lẻ, không thể chia đều
    if total_D % 2 != 0 or total_S % 2 != 0:
        with open('CAKE.OUT', 'w') as outfile:
            outfile.write("-1\n")
        return

    # Tổng số quả cần có ở mỗi phần
    half_D = total_D // 2
    half_S = total_S // 2

    # Duyệt qua các vị trí để tìm cách cắt
    for start in range(N):
        count_D = 0
        count_S = 0
        for end in range(start, start + N):  # Duyệt vòng tròn
            if cake[end % N] == 'D':
                count_D += 1
            else:
                count_S += 1

            # Kiểm tra điều kiện
            if count_D == half_D and count_S == half_S:
                a = (start + 1)  # Thêm 1 vì yêu cầu số nguyên dương
                b = (end % N + 1)  # Thêm 1 vì yêu cầu số nguyên dương
                if a > b:
                    a, b = b, a  # Đảm bảo a < b
                with open('CAKE.OUT', 'w') as outfile:
                    outfile.write(f"{a} {b}\n")
                return

    # Nếu không tìm thấy cách chia
    with open('CAKE.OUT', 'w') as outfile:
        outfile.write("-1\n")

if __name__ == '__main__':
    main()

File: THCS/2023_2024/TuyenHoa_2324/Cau3/CAKE.cpp
#include <iostream>
#include <fstream>
#include <string>

int main() {
    // Đọc dữ liệu từ file CAKE.INP
    std::ifstream infile("CAKE.INP");
    int N;
    infile >> N;
    std::string cake;
    infile >> cake;
    infile.close();

    // Biến lưu tổng số quả dâu và sim
    int totalD = 0, totalS = 0;

    // Tính tổng số quả dâu và sim
    for (char c : cake) {
        if (c == 'D') totalD++;
        else if (c == 'S') totalS++;
    }

    // Nếu tổng số quả dâu hoặc sim là lẻ thì không thể chia đều
    if (totalD % 2 != 0 || totalS % 2 != 0) {
        std::ofstream outfile("CAKE.OUT");
        outfile << -1 << std::endl;
        outfile.close();
        return 0;
    }

    // Biến lưu số lượng quả dâu và sim trong đoạn hiện tại
    int countD, countS;

    // Duyệt qua từng vị trí để tìm cách chia
    for (int start = 0; start < N; start++) {
        countD = 0;
        countS = 0;

        // Tính số lượng quả trong đoạn từ start đến start + N/2
        for (int length = 0; length < N; length++) {
            if (cake[(start + length) % N] == 'D') {
                countD++;
            } else {
                countS++;
            }

            // Kiểm tra nếu số lượng quả đến hiện tại đã đủ chia đều
            if (length + 1 >= N / 2) {
                // Kiểm tra điều kiện cắt bánh
                if (countD == totalD / 2 && countS == totalS / 2) {
                    int end = (start + length) % N;

                    // Định dạng kết quả a, b (a < b)
                    int a = start + 1;
                    int b = end + 1;

                    // Điều chỉnh b nếu b < a
                    if (b <= a) {
                        b += N;
                    }

                    std::ofstream outfile("CAKE.OUT");
                    outfile << a << " " << b << std::endl; // Ghi kết quả
                    outfile.close();
                    return 0; // Kết thúc chương trình
                }
            }
        }
    }

    // Nếu không tìm thấy cách chia, ghi -1 vào file
    std::ofstream outfile("CAKE.OUT");
    outfile << -1 << std::endl;
    outfile.close();

    return 0;
}
